File: composite-builds/temple-gradle-plugin/src/main/kotlin/com/jamesfchen/Task.java
package com.jamesfchen;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Target(ElementType.TYPE)
@Retention(RetentionPolicy.CLASS)
public @interface Task {
    InitPhase initPhase() default InitPhase.onCreate;
    InitProcess initProcess() default InitProcess.Main;
    Priority initPriority() default Priority.Normal;

    enum InitPhase {
        attachBaseContext,
        onCreate,
        onIdle,
    }

    enum InitProcess {
        All,
        Main,
        Launcher,
        Other
    }

    enum Priority {
        High,
        Normal,
        Low
    }
}

File: android/main-module/src/main/java/com/jamesfchen/Task.kt
package com.jamesfchen

@Target(AnnotationTarget.ANNOTATION_CLASS, AnnotationTarget.CLASS)
@kotlin.annotation.Retention(AnnotationRetention.BINARY)
annotation class Task(
    val initPhase: InitPhase = InitPhase.onCreate,
    val initProcess: InitProcess = InitProcess.Main,
    val initPriority: Priority = Priority.Normal
) {
    enum class InitPhase {
        attachBaseContext, onCreate, onIdle
    }

    enum class InitProcess {
        All, Main, Launcher, Other
    }

    enum class Priority {
        High, Normal, Low
    }
}